### PR TITLE
Logging: Add log decoration

### LINKED ISSUE #988: Add log decoration support
_Duplicate of [LOGGING-7303: Local Log Decoration (Internal)](https://newrelic.atlassian.net/browse/LOGGING-7303)_

Specifications described in [Agent Logs-in-Context Decoration from Engineering Brief: Logging in Agents (Internal)](https://docs.google.com/document/d/1HG9RD0NAYiLtR8ZFfH0obLbVZiVmQxOt3k8xBz-RZaQ/edit#heading=h.xgjthpm746nl)

Consider **Option A: Direct Injection** as the selected strategy.  This involves appending the following onto the log line itself: `NR-LINKING | {entity.guid} | {hostname} | {trace.id} | {span.id} |` if the setting is enabled.

Adding the config option to enable local log decoration should be completed as part of this ticket if it is not already complete with #983 

Log decoration should be allowed in high security mode and on by default.

As this is getting completed after #978, adding functionality to disable log decoration when `application_logging.enabled` is false is also within the scope of this issue.

### PR DESCRIPTION
**Pull request by kaylareopelle**
# Overview
Add local log decoration with linking metadata when `application_logging.local_decoration.enabled` is true.

# Related Github Issue
Closes #988 

# Testing
* Adds unit tests for new class, LocalLogDecorator
* Adds integration tests to Logger instrumentation
* Adds performance tests for log decorating

### CODE REVIEW

File: lib/new_relic/agent/local_log_decorator.rb
Comment (kaylareopelle) [inline lib/new_relic/agent/local_log_decorator.rb:22]: Commenting out until PR #1007 is merged

File: lib/new_relic/agent/instrumentation/logger/instrumentation.rb
Comment (kaylareopelle) [inline lib/new_relic/agent/instrumentation/logger/instrumentation.rb:42]: I'm a little hesitant about putting the instrumentation here. I keep going back and forth with the `LogEventAggregator`. Though it's a rare case, if someone has forwarding and decoration enabled, and the message is longer than 32 kb, the decoration would be truncated. If it's in the LogEventAggregator, a lot of extra code is run to add decoration that may not be necessary. 

One idea could be to add some complexity to the truncation, where it finds the size of the log decoration, truncates the message for decoration_size - max bytes and re-applies the decoration with the truncated message. This seems like it might be too much overhead for each logging line.

Thoughts?

File: lib/new_relic/agent/local_log_decorator.rb
Comment (kaylareopelle) [inline lib/new_relic/agent/local_log_decorator.rb:22]: PR #1007 has not been merged yet, so the setting doesn't exist. I plan to add this back once that PR is merged, unless anyone has objections!

File: lib/new_relic/agent/instrumentation/logger/instrumentation.rb
Comment (jasonrclark) [reply lib/new_relic/agent/instrumentation/logger/instrumentation.rb:42]: So my perspective here -- I'd say that we should be doing it in the instrumentation vs the aggregator... but I think we should do it _after_ we've done record.

If we're sending from the agent, we don't need the linking in the message as it'll be tacked on with the collector post. We also don't want that our truncating of what we're sending up to NR ends up truncating their local output, as they may care about that being complete.

**Review by fallwith — APPROVED** [review]
👍

**Comment by kaylareopelle** [comment]
Blocked by #1007

**Review by tannalynn — APPROVED** [review]
Looks good! 👍

File: lib/new_relic/agent/instrumentation/logger/instrumentation.rb
Comment (fallwith) [inline lib/new_relic/agent/instrumentation/logger/instrumentation.rb:50]: 😎

**Pull request merged**
